### PR TITLE
feat: add `--verbose` to demo/verbose

### PR DESCRIPTION
**Pull request by DarthBenro008**
# TL;DR

Adds `--verbose` command for the following commands:

- `flytectl sandbox start`
- `flytectl sandbox teardown`
- `flytectl demo start`
- `flytectl demo teardown`

## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [x] Code completed
- [x] Smoke tested
- [ ] Unit tests added
- [x] Code documentation added
- [x] Any pending items have an associated Issue

### For `start` command

Pretty straight forward, it uses a `client-go` package. Hence queried the event in the `flyte` namespace to show the logs

image

### For the `teardown` command

- This option did not have flags, hence had to add the flag command
- This command uses Docker instead of k8s, hence had to add the `Events` in interface

image

## Tracking Issue
- https://github.com/flyteorg/flyte/issues/2491

## Follow-up issue
_NA_

### CODE REVIEW

File: cmd/demo/teardown.go [hunk 1]
```diff
@@ -24,9 +25,10 @@ Usage
 )
 
 func teardownDemoCluster(ctx context.Context, args []string, cmdCtx cmdCore.CommandContext) error {
+	sandboxDefaultConfig := sandboxCmdConfig.DefaultConfig
 	cli, err := docker.GetDockerClient()
 	if err != nil {
 		return err
 	}
-	return sandbox.Teardown(ctx, cli)
+	return sandbox.Teardown(ctx, cli, sandboxDefaultConfig.Verbose)
```
Comment (pmahindrakar-oss) [inline cmd/demo/teardown.go:33]: I wasn't able to use this config for teardown. I think you are missing the config flag that ties up the command and its default config

```
| Normal | Created           | Created container kubernetes-dashboard                                                           |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Started           | Started container kubernetes-dashboard                                                           |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Pulled            | Successfully pulled image "ecr.flyte.org/bitnami/minio:2021.10.13-debian-10-r0" in 38.115757152s |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Created           | Created container minio                                                                          |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Started           | Started container minio                                                                          |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Pulled            | Successfully pulled image "ecr.flyte.org/ubuntu/postgres:13-21.04_beta" in 1m1.123476844s        |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Created           | Created container postgres                                                                       |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
| Normal | Started           | Started container postgres                                                                       |
+--------+-------------------+--------------------------------------------------------------------------------------------------+
👨‍💻 Flyte is ready! Flyte UI is available at http://localhost:30080/console 🚀 🚀 🎉 
Add KUBECONFIG and FLYTECTL_CONFIG to your environment variable 
export KUBECONFIG=$KUBECONFIG:/Users/praful/.kube/config:/Users/praful/.flyte/k3s/k3s.yaml 
export FLYTECTL_CONFIG=/Users/praful/.flyte/config-sandbox.yaml 
➜  flytectl git:(flyte/#2491) ✗ flytectl demo teardown --verbose
Error: unknown flag: --verbose
ERRO[0000] unknown flag: --verbose  
```

File: coverage.txt [hunk 2]
```diff
@@ -0,0 +1,2440 @@
+mode: atomic
```
Comment (pmahindrakar-oss) [inline coverage.txt:1]: Please remove this file

File: pkg/sandbox/start.go [hunk 3]
```diff
@@ -64,6 +64,31 @@ func isPodReady(v corev1api.Pod) bool {
...
+	table.SetAutoWrapText(false)
+	table.SetAutoFormatHeaders(true)
+
+	events, err := client.Events(flyteNamespace).List(ctx, v1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot get verbose logs for the deployment")
+	}
+	if len(events.Items) > 0 {
+		for _, event := range events.Items {
+			table.Append([]string{event.Type, event.Reason, event.Message})
```
Comment (pmahindrakar-oss) [inline pkg/sandbox/start.go:83]: Can we also add event.InvolvedObject.Name instead of Type.

File: pkg/sandbox/start.go [hunk 4]
```diff
@@ -269,6 +294,13 @@ func StartCluster(ctx context.Context, args []string, sandboxConfig *sandboxCmdC
 		if err := WatchFlyteDeployment(ctx, k8sClient.CoreV1()); err != nil {
 			return err
 		}
+
+		if sandboxConfig.Verbose {
+			if err = getEventLogs(ctx, k8sClient.CoreV1()); err != nil {
+				return err
+			}
+		}
```
Comment (pmahindrakar-oss) [inline pkg/sandbox/start.go:302]: nit : Can we Stream these while call  watch flyte deployment and also be available once the watch finishes.

**Review by pmahindrakar-oss — COMMENTED** [review]
Thanks for adding these. 
Also wanted to catch more verbose logging for other components sandbox and demo start. eg docker daemon , k3s server, helm install etc. Your changes should cover for helm install portion but if anything fails before that we wanted to have more verbose logging for those aswell
https://github.com/flyteorg/flyte/blob/master/docker/sandbox-lite/flyte-entrypoint-dind.sh

Can we see if those could be included aswell as part of this ticket

**Comment by DarthBenro008** [comment]
@pmahindrakar-oss Thank you for your review! 😃  I have made the required changes. The issue had stated with verbose logging of `helm` and `kubectl` commands which this PR very well covers. 

This PR also introduces the framework to listen for Docker events:
https://github.com/DarthBenro008/flytectl/blob/6718ab1942599078c52ad6f1ac73f08c4e645cf4/pkg/docker/docker.go#L29

I think for verbose logging of other components, its better to raise another issue (I would be happy to) and work on it, as this would keep things decoupled, less cluttered and easier to track/revert in future. 

Please let me know your thoughts.

**Comment by pmahindrakar-oss** [comment]
Thanks @DarthBenro008 . Sounds good. Would you mind opening the follow up issue to capture the rest of the components logs aswell.

**Review by pmahindrakar-oss — APPROVED** [review]

**Comment by samhita-alla** [comment]
@DarthBenro008 @pmahindrakar-oss, one of the checks is failing: https://github.com/flyteorg/flytectl/pull/359#pullrequestreview-1134262787.

**Comment by pmahindrakar-oss** [comment]
@DarthBenro008 can you revert the change to generated enumer files which is probably causing the generate checks to fail

**Comment by DarthBenro008** [comment]
@pmahindrakar-oss done!

**Comment by pmahindrakar-oss** [comment]
Thanks @DarthBenro008 . Minor UX issues with table rendering on teardown

```
 flytectl demo teardown --verbose

---- Verbose Logs ----
+----+------+------+--------+
| ID | FROM | TYPE | ACTION |
+----+------+------+--------+
| 383ccd2e48b000d2552d382bedbebd101362b05ceac9909743d324390e1519cb | sha256:dca9be805809cc6337d20565346fa25de29b250a352b3ce2431f279f47b7320d | container | kill   |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+--------+
| 383ccd2e48b000d2552d382bedbebd101362b05ceac9909743d324390e1519cb | sha256:dca9be805809cc6337d20565346fa25de29b250a352b3ce2431f279f47b7320d | container | die    |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+--------+
|                                                                  |                                                                         | volume    | unmount |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+---------+
|                                                                  |                                                                         | volume    | unmount |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+---------+
|                                                                  |                                                                         | volume    | unmount |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+---------+
|                                                                  |                                                                         | volume    | unmount |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+---------+
|                                                                  |                                                                         | volume    | unmount |
+------------------------------------------------------------------+-------------------------------------------------------------------------+-----------+---------+
context removed for "flyte-sandbox".
🧹 🧹 Sandbox cluster is removed successfully.
```

File: pkg/sandbox/teardown.go [hunk 5]
```diff
@@ -33,6 +51,24 @@ func Teardown(ctx context.Context, cli docker.Docker) error {
...
+	table.SetAutoWrapText(true)
+	table.SetAutoFormatHeaders(true)
+	msgs, _ := cli.Events(ctx, types.EventsOptions{
+		Since: "1m",
+	})
+	table.Render()
+	for m := range msgs {
+		table.RenderRowOnce([]string{m.ID, m.From, m.Type, m.Action})
+	}
+}
```
Comment (DarthBenro008) [inline pkg/sandbox/teardown.go:70]: @pmahindrakar-oss as this is a `chan` which is run by a goroutine, i have no other way to render upcoming messages (other than rendering the whole table again when new message comes, which will print multiple tables). Hence i was forced to use the `RenderRowOnce` function instead, the caveat being the title's won't adjust themselves to the size. 

I was not able to find a workaround for this, hence the UX issue. Any suggestions would be helpful

Comment (pmahindrakar-oss) [reply pkg/sandbox/teardown.go:70]: You can use consume the message in the same go routine and  print in the main routine. you can use wait groups for this. And also would be preferable if you utilize the pkg/printer utility

Comment (samhita-alla) [reply pkg/sandbox/teardown.go:70]: @DarthBenro008, would you be able to work on this?
